### PR TITLE
[streaming] don't fail on EOF errors, use stream as pointer

### PR DESCRIPTION
**Pull request by azr**
Hi,
tldr, what this pr does ?

* Using `*Stream` instead of `Stream` for a stream has a state ! ( `runing`, it's actually a bool right now)
 * this simplifies greatly the termination of a streaming by just calling stream.Stop()
* fix when stream socket get's closed, reopen it instead of failing.
* fix some failing tests ( had to change import path for this ). 
* added twitter timed backoffs
  * twitter defines every backoff scenario and how we should play with them, I made handlers that return them.
  *  I applied the too many requests backoff to the stream section
* add automatic testing using a travis.yml file
  * For it to work on this repo you will need to click a `+` button on travis' [website](https://travis-ci.org/) and set your twitter acces tokens (don't forget to hide them), then it will test all your prs for you ! :D.
  * it's all green : https://travis-ci.org/azr/anaconda

Thanks :)

### CODE REVIEW

**Comment by azr** [comment]
Hello there, I just rebased this pr !
Sorry I didn't have the time to comment it when I made it :).
May be you want me to split the commits into different prs so you can see more clearly ?

I'll explain it here anyways.

### EOF
This adds a " bugfix " in cases when twitter sends us an 'EOF' token right after we open the stream session.

It is a rare case but happens when the internet connection is feeling bad.

It also feels like sometimes twitter uses it as an 'Enhance your calm' fallback.

**Comment by dimroc** [comment]
+1

**Comment by azr** [comment]
Hello there, this pr completely changed, it only fixes EOF errors and starts using streams as pointers

**Comment by azr** [comment]
Hello there,

please tell me what's wrong !!

In the meantime we had a look at the doc from twitter, and they define some backoff policies, which I started using for the streaming api.

I made some helpers so other people can use them too.

o/

**Comment by azr** [comment]
I am sorry about spamming this pr, I'll make a new one more clear if no one comments on this.

I am also sorry if I made something wrong, please tell me if it's the case.

In the meantime I applied twitter's backoff policies and made some helpers returning backoff interfaces so anyone can use them anywhere.

Cheers !

**Comment by azr** [comment]
I also added a .travis.yml file so we know if the projects tests fails, which is the case !

```
./twitter_test.go:142: undefined: anaconda.EntityMedia
./twitter_test.go:172: multiple-value api.GetFriendsIdsAll() in single-value context
./twitter_test.go:230: undefined: anaconda.TwitterErrorDoesNotExist2
```

**Comment by azr** [comment]
This has nothing to do with this pr but I'll try to fix them.

**Comment by azr** [comment]
Here, it's all fixed: https://travis-ci.org/azr/anaconda 

You can add env vars to a travis build.

ps: if multiple builds are run at the same time you will get this kind of errors: 
`
--- FAIL: Test_DMScreenName (29.94s)
	twitter_test.go:296: Get https://api.twitter.com/1.1/direct_messages/new.json returned status 403, {"errors":[{"code":151,"message":"There was an error sending your message: Whoops! You already said that."}]}
`
> https://travis-ci.org/azr/anaconda/jobs/72873081

**Comment by azr** [comment]
bump

**Comment by azr** [comment]
This pr has been running in production for 9 months now !

**Comment by ChimeraCoder** [comment]
Sorry about that! This is now merged as of 4b2bf11c3f29812ce4507e50d43864e90f0.

Thanks for the pull request, and for testing this!

**Comment by ChimeraCoder** [comment]
(marked as closed, but really this is merged, because Github doesn't understand how offline merges work)

**Comment by azr** [comment]
cool :)

**Comment by azr** [comment]
Don't forget to setup https://travis-ci.org/chimeracoder/anaconda  ( if you're okay with this :) )

**Comment by ChimeraCoder** [comment]
@azr Thanks! It's up at https://travis-ci.org/ChimeraCoder/anaconda (case-sensitive URL).

It took a bit of time because I had to fix the tests so that they would work from mock HTTP responses; otherwise, the tests would be unreliable due to Twitter's throttling (depending on how many CI builds Travis ran in parallel).

**Comment by azr** [comment]
Awesome !! :)
